File: algo/sudoku_solver.c
#include <stdio.h>
#include <stdbool.h>
#include <time.h>

#define SIZE 9

bool isValid(int board[SIZE][SIZE], int row, int col, int ch) {
    // Check the row
    for (int i = 0; i < SIZE; i++) {
        if (board[row][i] == ch) return false;
    }

    // Check the column
    for (int i = 0; i < SIZE; i++) {
        if (board[i][col] == ch) return false;
    }

    // Check the 3x3 sub-grid
    int startRow = (row / 3) * 3;
    int startCol = (col / 3) * 3;
    for (int i = startRow; i < startRow + 3; i++) {
        for (int j = startCol; j < startCol + 3; j++) {
            if (board[i][j] == ch) return false;
        }
    }

    return true;
}

bool solve(int board[SIZE][SIZE]) {
    for (int i = 0; i < SIZE; i++) {
        for (int j = 0; j < SIZE; j++) {
            if (board[i][j] == 0) { // Find an empty cell
                for (int ch = 1; ch <= 9; ch++) { // Try numbers 1 to 9
                    if (isValid(board, i, j, ch)) {
                        board[i][j] = ch; // Assign the number
                        if (solve(board)) { // Recursively solve the rest
                            return true;
                        }
                        board[i][j] = 0; // Undo the assignment if not valid
                    }
                }
                return false; // If no valid number found, return false
            }
        }
    }
    return true; // Puzzle solved
}

void printBoard(int board[SIZE][SIZE]) {
    for (int i = 0; i < SIZE; i++) {
        for (int j = 0; j < SIZE; j++) {
            printf("%d ", board[i][j]);
        }
        printf("\n");
    }
}

int main() {
    // Array of puzzles
    int puzzles[9][SIZE][SIZE] = {
        {
            {5, 3, 0, 0, 7, 0, 0, 0, 0},
            {6, 0, 0, 1, 9, 5, 0, 0, 0},
            {0, 9, 8, 0, 0, 0, 0, 6, 0},
            {8, 0, 0, 0, 6, 0, 0, 0, 3},
            {4, 0, 0, 8, 0, 3, 0, 0, 1},
            {7, 0, 0, 0, 2, 0, 0, 0, 6},
            {0, 6, 0, 0, 0, 0, 2, 8, 0},
            {0, 0, 0, 4, 1, 9, 0, 0, 5},
            {0, 0, 0, 0, 8, 0, 0, 7, 9}
        },
        {
            {0, 2, 0, 6, 0, 8, 0, 0, 0},
            {5, 8, 0, 0, 0, 9, 7, 0, 0},
            {0, 0, 0, 0, 4, 0, 0, 0, 0},
            {3, 7, 0, 0, 0, 0, 5, 0, 0},
            {6, 0, 0, 0, 0, 0, 0, 0, 4},
            {0, 0, 8, 0, 0, 0, 0, 1, 3},
            {0, 0, 0, 0, 2, 0, 0, 0, 0},
            {0, 0, 9, 8, 0, 0, 0, 3, 6},
            {0, 0, 0, 3, 0, 6, 0, 9, 0}
        },
        {
            {0, 0, 0, 0, 0, 0, 0, 1, 2},
            {0, 0, 0, 0, 3, 5, 0, 0, 0},
            {0, 0, 0, 4, 0, 0, 0, 0, 0},
            {0, 0, 8, 0, 0, 0, 0, 0, 0},
            {0, 0, 0, 5, 0, 1, 0, 0, 0},
            {0, 0, 0, 0, 0, 0, 6, 0, 0},
            {0, 0, 0, 0, 0, 4, 0, 0, 0},
            {0, 0, 0, 9, 8, 0, 0, 0, 0},
            {9, 5, 0, 0, 0, 0, 0, 0, 0}
        },
        {
            {0, 0, 0, 6, 0, 0, 0, 0, 0},
            {0, 0, 0, 0, 0, 0, 4, 0, 0},
            {0, 0, 0, 0, 0, 0, 0, 0, 3},
            {0, 0, 7, 0, 0, 9, 0, 2, 0},
            {0, 0, 0, 0, 4, 0, 0, 0, 0},
            {0, 5, 0, 7, 0, 0, 8, 0, 0},
            {1, 0, 0, 0, 0, 0, 0, 0, 0},
            {0, 0, 3, 0, 0, 0, 0, 0, 0},
            {0, 0, 0, 0, 0, 0, 2, 0, 0}
        },
        {
            {0, 0, 0, 8, 0, 0, 0, 0, 0},
            {0, 0, 0, 0, 0, 0, 0, 6, 0},
            {0, 0, 0, 0, 0, 0, 0, 0, 4},
            {0, 3, 0, 0, 5, 0, 0, 0, 0},
            {0, 0, 0, 7, 0, 8, 0, 0, 0},
            {0, 0, 0, 0, 6, 0, 0, 2, 0},
            {5, 0, 0, 0, 0, 0, 0, 0, 0},
            {0, 2, 0, 0, 0, 0, 0, 0, 0},
            {0, 0, 0, 0, 0, 4, 0, 0, 0}
        },
        {
            {0, 0, 0, 0, 0, 0, 2, 0, 0},
            {0, 0, 0, 0, 3, 0, 0, 0, 9},
            {0, 0, 0, 7, 0, 0, 0, 0, 0},
            {0, 0, 0, 0, 0, 9, 0, 0, 0},
            {0, 0, 4, 0, 0, 0, 3, 0, 0},
            {0, 9, 0, 0, 0, 0, 0, 0, 0},
            {0, 0, 0, 0, 0, 0, 0, 8, 0},
            {0, 5, 0, 0, 0, 0, 0, 0, 0},
            {0, 0, 2, 0, 0, 0, 0, 0, 0}
        },
        {
            {0, 0, 0, 0, 0, 0, 0, 1, 0},
            {0, 0, 0, 0, 6, 0, 0, 0, 0},
            {0, 0, 0, 0, 0, 8, 0, 0, 0},
            {0, 4, 0, 0, 0, 0, 0, 0, 3},
            {0, 0, 0, 0, 7, 0, 0, 0, 0},
            {9, 0, 0, 0, 0, 0, 0, 6, 0},
            {0, 0, 0, 5, 0, 0, 0, 0, 0},
            {0, 0, 1, 0, 0, 0, 0, 0, 0},
            {0, 8, 0, 0, 0, 0, 0, 0, 0}
        },
        {
            {0, 0, 0, 0, 0, 0, 0, 0, 1},
            {0, 0, 0, 0, 0, 2, 0, 0, 0},
            {0, 0, 0, 3, 0, 0, 0, 0, 0},
            {0, 0, 7, 0, 0, 0, 0, 0, 0},
            {0, 6, 0, 0, 0, 0, 0, 0, 0},
            {0, 0, 0, 0, 0, 5, 0, 0, 0},
            {0, 0, 0, 0, 0, 0, 0, 0, 0},
            {1, 0, 0, 0, 0, 0, 0, 0, 0},
            {0, 0, 0, 0, 4, 0, 0, 0, 0}
        }
    };

    for (int puzzleIndex = 0; puzzleIndex < 9; puzzleIndex++) {
        int board[SIZE][SIZE];
        for (int i = 0; i < SIZE; i++) {
            for (int j = 0; j < SIZE; j++) {
                board[i][j] = puzzles[puzzleIndex][i][j];
            }
        }

        // Start timing
        clock_t start = clock();

        // Solve the puzzle
        bool solved = solve(board);

        // End timing
        clock_t end = clock();
        double duration = ((double) (end - start)) / CLOCKS_PER_SEC;

        // Output the result and time taken
        printf("Puzzle %d %s\n", puzzleIndex + 1, solved ? "solved!" : "not solvable");
        printf("Time taken: %.6f seconds\n", duration);

        // Uncomment this to print the solved board
        // if (solved) {
        //     printBoard(board);
        // }
        printf("-----------------------------\n");
    }

    return 0;
}

File: algo/sudoku_solver.go
package main

import (
	"fmt"
	"sync"
	"time"
)

const SIZE = 9

// Utility functions to print the Sudoku board
func printBoard(board [][]int) {
	for i := 0; i < SIZE; i++ {
		for j := 0; j < SIZE; j++ {
			fmt.Printf("%d ", board[i][j])
		}
		fmt.Println()
	}
}

// Checks whether placing 'num' at position (row, col) is valid
func isValid(board [][]int, row, col, num int) bool {
	// Check the row
	for i := 0; i < SIZE; i++ {
		if board[row][i] == num {
			return false
		}
	}

	// Check the column
	for i := 0; i < SIZE; i++ {
		if board[i][col] == num {
			return false
		}
	}

	// Check the 3x3 subgrid
	startRow := row - row%3
	startCol := col - col%3
	for i := 0; i < 3; i++ {
		for j := 0; j < 3; j++ {
			if board[startRow+i][startCol+j] == num {
				return false
			}
		}
	}
	return true
}

// Solves the Sudoku using backtracking and goroutines for parallelization
func solve(board [][]int, wg *sync.WaitGroup) bool {
	defer wg.Done() // Decrement the counter when the function finishes

	// Find the next empty cell
	for row := 0; row < SIZE; row++ {
		for col := 0; col < SIZE; col++ {
			if board[row][col] == 0 {
				// Try every number from 1 to 9
				for num := 1; num <= SIZE; num++ {
					if isValid(board, row, col, num) {
						board[row][col] = num

						// Create a new WaitGroup for the next level of recursion
						var innerWg sync.WaitGroup
						innerWg.Add(1)
						go func() {
							if solve(board, &innerWg) {
								wg.Done() // Notify the outer WaitGroup
								return
							}
							board[row][col] = 0
							innerWg.Done() // Mark this recursive branch as done
						}()
						return false
					}
				}
				return false // No valid number found, need to backtrack
			}
		}
	}
	return true // Puzzle solved
}

func main() {
	// Define a Sudoku puzzle (you can modify it for testing)
	board := [][]int{
		{5, 3, 0, 0, 7, 0, 0, 0, 0},
		{6, 0, 0, 1, 9, 5, 0, 0, 0},
		{0, 9, 8, 0, 0, 0, 0, 6, 0},
		{8, 0, 0, 0, 6, 0, 0, 0, 3},
		{4, 0, 0, 8, 0, 3, 0, 0, 1},
		{7, 0, 0, 0, 2, 0, 0, 0, 6},
		{0, 6, 0, 0, 0, 0, 2, 8, 0},
		{0, 0, 0, 4, 1, 9, 0, 0, 5},
		{0, 0, 0, 0, 8, 0, 0, 7, 9},
	}

	var wg sync.WaitGroup
	wg.Add(1)

	// Start timing
	start := time.Now()

	// Solve the Sudoku puzzle
	if solve(board, &wg) {
		fmt.Println("Puzzle solved:")
		printBoard(board)
	} else {
		fmt.Println("Puzzle could not be solved")
	}

	// Wait for all goroutines to finish
	wg.Wait()

	// End timing
	duration := time.Since(start)
	fmt.Println("Time taken:", duration.Seconds(), "seconds")
}
